changelogs/fragments/50882-docker_swarm_service-fix-publish-idempotency.yml

```
@@ -0,0 +1,2 @@
+bugfixes:
+- "docker_swarm_service - fixing falsely reporting ``publish`` as changed when ``publish.mode`` is not set."
```

```suggestion
- "docker_swarm_service - fixing falsely reporting ``publish`` as changed when ``publish.mode`` or ``publish.protocol`` is not set, or when the order changed."
```
Not setting `publish.protocol` should have had the same effect (I think), as (until now) it was compared to what Docker returned, and that contained `protocol: tcp` if it wasn't specified.

BTW, I think adding a test which switches the order (to see whether that's idempotent) is probably a good idea, too. Changing the first idempotent test (just switch the elements in the `publish:` list) should be enough for this.

---

Hmm I don’t think that’s the case since it always defaulted to tcp in `from_ansible_params` when not set:
`param_p.get('protocol', 'tcp')`

Yes that’s a good test, I’ll add it.

---

Yep, you're right.